### PR TITLE
Replaces empty strings `""` in `prepare_matches_mapper` dataframe with `NA` values

### LINKED ISSUE #66: How can we get non-missing outputs of `matched_*` and `unit`?
`prepare_****_product()` with the example datasets yield only missing in columns `matched_*` and `unit`. I see this in:

* `prepare_pctr_product()`
* `prepare_pstr_product()`
* `prepare_istr_product()`
* `prepare_ictr_product()`

Questions:

1. Is this expected? 
2. How can we get-non missing values, and how can we test that behaviour?
3. Note we represent missingness in two ways: with `NA` and with the empty string `""`. Is this expected?

``` r
library(dplyr, warn.conflicts = FALSE)
library(tiltIndicatorAfter)

emissions_profile_at_product_level <- prepare_pctr_product(
  pctr_product,
  ep_companies,
  ecoinvent_activities,
  matches_mapper,
  isic_tilt_mapper
)

emissions_profile_at_product_level |> 
  glimpse()
#> Rows: 25
#> Columns: 21
#> $ companies_id                       <chr> "id3", "id1", "id1", "id1", "id1", …
#> $ company_name                       <chr> "company C", "company A", "company …
#> $ country                            <chr> "austria", "germany", "germany", "g…
#> $ PCTR_risk_category                 <chr> NA, "high", "medium", "low", "high"…
#> $ benchmark                          <chr> NA, "all", "isic_sec", "tilt_sec", …
#> $ ep_product                         <chr> NA, "building construction", "build…
#> $ matched_activity_name              <chr> NA, "", "", "", "", "", "", "", "",…
#> $ matched_reference_product          <chr> NA, "", "", "", "", "", "", "", "",…
#> $ unit                               <chr> NA, "", "", "", "", "", "", "", "",…
#> $ multi_match                        <chr> NA, "FALSE", "FALSE", "FALSE", "FAL…
#> $ matching_certainty                 <chr> NA, "high", "high", "high", "high",…
#> $ matching_certainty_company_average <chr> NA, "medium", "medium", "medium", "…
#> $ tilt_sector                        <chr> NA, "tilt_sector A", "tilt_sector A…
#> $ tilt_subsector                     <chr> NA, "tilt_subsector A", "tilt_subse…
#> $ isic_4digit                        <chr> NA, "3020", "3020", "3020", "3020",…
#> $ isic_name                          <chr> NA, "Manufacture of railway locomot…
#> $ company_city                       <chr> "voitsberg", "frankfurt", "frankfur…
#> $ postcode                           <dbl> 8570, 60316, 60316, 60316, 60316, 6…
#> $ address                            <chr> "ruhmannstrasse 1 | 8570 voitsberg"…
#> $ main_activity                      <chr> "service provider", "manufacturer/ …
#> $ activity_uuid_product_uuid         <chr> NA, "ebd4dddf-9f74-5fd1-89ce-197b60…

emissions_profile_at_product_level |> count(matched_activity_name)
#> # A tibble: 2 × 2
#>   matched_activity_name     n
#>   <chr>                 <int>
#> 1 ""                       24
#> 2 <NA>                      1

emissions_profile_at_product_level |> count(matched_reference_product)
#> # A tibble: 2 × 2
#>   matched_reference_product     n
#>   <chr>                     <int>
#> 1 ""                           24
#> 2 <NA>                          1

emissions_profile_at_product_level |> count(unit)
#> # A tibble: 2 × 2
#>   unit      n
#>   <chr> <int>
#> 1 ""       24
#> 2 <NA>      1
```

<sup>Created on 2023-09-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by kalashsinghal**
closes #66  
relates to #68 

@maurolepore This PR replaces empty strings `""` in `prepare_matches_mapper` dataframe with `NA` values. A toy dataset is being created to test this. Please review and let me know what else we can test with this toy dataset!

Note: Due to the left join between `matches_mapper` and `ecoinvent_activities` on `activity_uuid_product_uuid`, ids which are not present in `ecoinvent_activities` create null values for columns `reference_product_name`, `activity_name`, and `unit`. And ids which are present in both datasets have non-null values in all the other columns.

### CODE REVIEW

**Review by maurolepore — COMMENTED** [review]
Thanks Kalash. 

See my comments.

File: tests/testthat/test-prepare_matches_mapper.R [hunk 1]
```diff
@@ -7,3 +7,28 @@ test_that("`completion` column has either `low`, `medium`, or `high` values", {
...
+    category = c("b", "a", "all")
+  )
+
+  ecoinvent_activities <- tibble(
+    activity_uuid_product_uuid = c("any", NA_character_),
+    activity_name = c(NA_character_, "c"),
+    geography = c("d", NA_character_),
+    reference_product_name = c("d", NA_character_),
+    unit = c("d", "c")
+  )
```
Comment (maurolepore) [inline tests/testthat/test-prepare_matches_mapper.R:29]: The data seems to be a little larger than the minimum required. In general, tests with smaller data are easier to understand and thus maintain because the intent becomes clearer. 

My experiments suggest that we can test this with two 1-row dataframes where all values are `"a"` except for one value where `*uuid = NA_character`. This produces a failing test before we write the code, and a passing test after we write it.

We could make it easy to create such toy data with a little helper. For example:

``` r
# Helper
tibble_names <- function(x, nms) {
  m <- matrix(x, ncol = length(nms), dimnames = list(NULL, nms))
  tibble::as_tibble(m)
}
# Examples
tibble_names("a", names(cars))
#> # A tibble: 1 × 2
#>   speed dist 
#>   <chr> <chr>
#> 1 a     a
tibble_names(NA_character_, names(cars))
#> # A tibble: 1 × 2
#>   speed dist 
#>   <chr> <chr>
#> 1 <NA>  <NA>
tibble_names(c("a", NA_character_), names(cars))
#> # A tibble: 1 × 2
#>   speed dist 
#>   <chr> <chr>
#> 1 a     <NA>

# Test with `main` (to simulate the failing test before we write the code)
library(tiltIndicatorAfter)

# Arrange
mm <- tibble_names("a", names(matches_mapper))
mm$activity_uuid_product_uuid <- NA_character_
ea <- tibble_names("a", names(ecoinvent_activities))
# Act
out <- tiltIndicatorAfter:::prepare_matches_mapper(mm, ea)
# Assert
testthat::expect_equal(out$activity_uuid_product_uuid, NA_character_)
#> Error: out$activity_uuid_product_uuid (`actual`) not equal to NA_character_ (`expected`).
#> 
#> `actual`:   ""
#> `expected`: NA

# Test with this PR (to simulate the passing test after we write the code)
devtools::load_all()
#> ℹ Loading tiltIndicatorAfter

# Arrange
mm <- tibble_names("a", names(matches_mapper))
mm$activity_uuid_product_uuid <- NA_character_
ea <- tibble_names("a", names(ecoinvent_activities))
# Act
out <- tiltIndicatorAfter:::prepare_matches_mapper(mm, ea)
# Assert
testthat::expect_equal(out$activity_uuid_product_uuid, NA_character_)
```

<sup>Created on 2023-10-11 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/prepare_matches_mapper.R [hunk 2]
```diff
@@ -13,6 +13,7 @@ prepare_matches_mapper <- function(mapper, activities) {
     left_join(activities, by = "activity_uuid_product_uuid") |>
     summarise(across(everything(), ~ paste0(na.omit(unique(.)), collapse = "; ")), .by = c("country", "main_activity", "clustered")) |>
     ungroup() |>
+    mutate_all(~ifelse(. == "", NA_character_, .)) |>
     distinct() |>
     mutate(multi_match = case_when(
       multi_match == "" ~ "FALSE",
```
[the remark this answers is not in the document]
Comment (kalashsinghal) [reply R/prepare_matches_mapper.R:19]: @maurolepore Good catch!
Interestingly this test didn't fail: 
```test_that("`multi_match` column has either `TRUE` or `FALSE` values", {```
However, I have made the correction in next commit

File: NAMESPACE [hunk 3]
```diff
@@ -23,6 +23,7 @@ importFrom(dplyr,group_by)
 importFrom(dplyr,join_by)
 importFrom(dplyr,left_join)
 importFrom(dplyr,mutate)
+importFrom(dplyr,mutate_all)
```
[the remark this answers is not in the document]
Comment (kalashsinghal) [reply NAMESPACE:26]: changed

File: R/prepare_matches_mapper.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (kalashsinghal) [reply R/prepare_matches_mapper.R]: `NA`s in the `multi_match` column needs to be converted into `FALSE`. Hence I have made the correction accordingly.

File: tests/testthat/test-prepare_matches_mapper.R (hunk 1 above)
Comment (kalashsinghal) [reply tests/testthat/test-prepare_matches_mapper.R]: @maurolepore Yes you are right! The actual usecase can be tested with just 1-row dataframes. Thanks for giving the idea of using matrix! I have updated the test accordingly

File: R/tiltIndicatorAfter-package.R [hunk 4]
```diff
@@ -13,6 +13,7 @@
 #' @importFrom dplyr join_by
 #' @importFrom dplyr left_join
 #' @importFrom dplyr mutate
+#' @importFrom dplyr mutate_all
```
Comment (maurolepore) [inline R/tiltIndicatorAfter-package.R:16]: Are we still using `mutate_all()` anywhere? If not we should remove it here and regenerate NAMESPACE via `devtools::document()`.

File: tests/testthat/test-prepare_matches_mapper.R [hunk 5]
```diff
@@ -7,3 +7,31 @@ test_that("`completion` column has either `low`, `medium`, or `high` values", {
   out <- prepare_matches_mapper(small_matches_mapper, ecoinvent_activities)
   expect_equal(unique(out$completion), c("low", "high", "medium"))
 })
+
+test_that("`prepare_matches_mapper` doesn't have any empty string", {
+  tibble_names <- function(x, nms) {
+    m <- matrix(x, ncol = length(nms), dimnames = list(NULL, nms))
+    tibble::as_tibble(m)
+  }
```
Comment (maurolepore) [inline tests/testthat/test-prepare_matches_mapper.R:15]: This helper is already useful in 2 tests, and maybe more. So I think it's best to move it to R/utils-tests.R (or tests/testthat/helper.R but R/utils-tests.R is more discoverable).

File: tests/testthat/test-prepare_matches_mapper.R [hunk 6]
```diff
@@ -7,3 +7,31 @@ test_that("`completion` column has either `low`, `medium`, or `high` values", {
...
+
+  mm <- tibble_names("a", names(matches_mapper))
+  mm$multi_match <- NA_character_
+  ea <- tibble_names("a", names(ecoinvent_activities))
+
+  out <- prepare_matches_mapper(mm, ea)
+  expect_equal(unique(out$multi_match), c("FALSE"))
```
Comment (maurolepore) [inline tests/testthat/test-prepare_matches_mapper.R:36]: So we expect "FALSE" not `FALSE`?

``` r
devtools::load_all()
#> ℹ Loading tiltIndicatorAfter

tibble_names <- function(x, nms) {
  m <- matrix(x, ncol = length(nms), dimnames = list(NULL, nms))
  tibble::as_tibble(m)
}

mm <- tibble_names("a", names(matches_mapper))
mm$multi_match <- NA_character_
ea <- tibble_names("a", names(ecoinvent_activities))

out <- prepare_matches_mapper(mm, ea)

testthat::expect_false(out$multi_match)
#> Error: out$multi_match is not FALSE
#> 
#> `actual` is a character vector ('FALSE')
#> `expected` is a logical vector (FALSE)
```

<sup>Created on 2023-10-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/prepare_matches_mapper.R [hunk 7]
```diff
@@ -13,9 +13,10 @@ prepare_matches_mapper <- function(mapper, activities) {
     left_join(activities, by = "activity_uuid_product_uuid") |>
     summarise(across(everything(), ~ paste0(na.omit(unique(.)), collapse = "; ")), .by = c("country", "main_activity", "clustered")) |>
     ungroup() |>
+    mutate(across(everything(), ~ifelse(. == "", NA_character_, .))) |>
     distinct() |>
     mutate(multi_match = case_when(
-      multi_match == "" ~ "FALSE",
+      is.na(multi_match) ~ "FALSE",
```
Comment (maurolepore) [inline R/prepare_matches_mapper.R:19]: So we expect the character "FALSE" not the logical `FALSE`? See the test below.

**Review by maurolepore — COMMENTED** [review]
Hey @kalashsinghal, here is another round of revision. Please see my commments.

File: R/prepare_matches_mapper.R (hunk 7 above)
Comment (kalashsinghal) [reply R/prepare_matches_mapper.R:19]: @maurolepore this line of code:
```summarise(across(everything(), ~ paste0(na.omit(unique(.)), collapse = "; ")), .by = c("country", "main_activity", "clustered"))```
converts the `multi_match` column type from logical to character. However, there is no need to keep the column type to logical. Hence, I am choosing to convert it into character type. Please let me know if you think otherwise! :)

File: R/tiltIndicatorAfter-package.R (hunk 4 above)
Comment (kalashsinghal) [reply R/tiltIndicatorAfter-package.R:16]: resolved

File: tests/testthat/test-prepare_matches_mapper.R (hunk 6 above)
Comment (kalashsinghal) [reply tests/testthat/test-prepare_matches_mapper.R:36]: We expect "FALSE" and not logical FALSE

File: tests/testthat/test-prepare_matches_mapper.R (hunk 5 above)
Comment (kalashsinghal) [reply tests/testthat/test-prepare_matches_mapper.R]: Added to R/utils-tests.R

File: R/prepare_matches_mapper.R (hunk 7 above)
Comment (maurolepore) [reply R/prepare_matches_mapper.R:19]:
``` r
# Good
isTRUE(TRUE)
#> [1] TRUE
testthat::expect_true(TRUE)

# Confusing
isTRUE("TRUE")
#> [1] FALSE
testthat::expect_true("TRUE")
#> Error: "TRUE" is not TRUE
#> 
#> `actual` is a character vector ('TRUE')
#> `expected` is a logical vector (TRUE)
```

<sup>Created on 2023-10-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Comment (maurolepore) [reply R/prepare_matches_mapper.R:19]:
``` r
# Logicals allow you to do useful things
good <- tibble::tibble(x = c(TRUE, TRUE, FALSE))

good |> dplyr::filter(x)
#> # A tibble: 2 × 1
#>   x    
#>   <lgl>
#> 1 TRUE 
#> 2 TRUE

good |> dplyr::filter(!x)
#> # A tibble: 1 × 1
#>   x    
#>   <lgl>
#> 1 FALSE

sum(good$x)
#> [1] 2

mean(good$x)
#> [1] 0.6666667



# Characters are very differnt

bad <- tibble::tibble(x = c("TRUE", "TRUE", "FALSE"))

try(bad |> dplyr::filter(x))
#> Error in dplyr::filter(bad, x) : ℹ In argument: `x`.
#> Caused by error:
#> ! `..1` must be a logical vector, not a character vector.

try(bad |> dplyr::filter(!x))
#> Error in dplyr::filter(bad, !x) : ℹ In argument: `!x`.
#> Caused by error in `!x`:
#> ! invalid argument type

try(sum(bad$x))
#> Error in sum(bad$x) : invalid 'type' (character) of argument

try(mean(bad$x))
#> Warning in mean.default(bad$x): argument is not numeric or logical: returning
#> NA
#> [1] NA
```

<sup>Created on 2023-10-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Comment (kalashsinghal) [reply R/prepare_matches_mapper.R]: @maurolepore I now see a good reason to keep the column type to logical. Thanks for explaining its importance! Please review the new change! Thanks!

File: tests/testthat/test-prepare_matches_mapper.R (hunk 6 above)
Comment (kalashsinghal) [reply tests/testthat/test-prepare_matches_mapper.R]: @maurolepore I have changed it to `FALSE` (logical)

File: tests/testthat/test-prepare_matches_mapper.R [hunk 8]
```diff
@@ -1,9 +1,36 @@
...
+  mm <- tibble(
+    country = c("x", "y"),
+    main_activity = c("a", "a"),
+    clustered = c("a", "a"),
+    ep_id = c("a", "a"),
+    activity_uuid_product_uuid = c("a", "a"),
+    multi_match = c(NA_character_, ""),
+    completion = c("a", "a"),
+    category = c("a", "a")
+  )
```
Comment (maurolepore) [inline tests/testthat/test-prepare_matches_mapper.R:30]: This test doesn't seem correct. In the dataset `matches_mapper` the column `multi_match` is logical.

``` r
is.logical(tiltIndicatorAfter::matches_mapper$multi_match)
#> [1] TRUE
```

We could enforce this with an assertion, so we're sure it's never a character.

```r
prepare_matches_mapper <- function(mapper, activities) {
  stopifnot(is.logical(mapper$multi_match))
  ...
}
```

Then we don't need this line:

```r
   ...
    mutate(multi_match = as.logical(.data$multi_match)) |>
```

But I'm getting a bit lost and may have lost track of how we arrived here :-)

File: R/prepare_matches_mapper.R [hunk 9]
```diff
@@ -12,10 +12,11 @@ prepare_matches_mapper <- function(mapper, activities) {
   mapper |>
     left_join(activities, by = "activity_uuid_product_uuid") |>
     summarise(across(everything(), ~ paste0(na.omit(unique(.)), collapse = "; ")), .by = c("country", "main_activity", "clustered")) |>
-    ungroup() |>
+    mutate(across(everything(), ~ifelse(. == "", NA_character_, .))) |>
     distinct() |>
+    mutate(multi_match = as.logical(.data$multi_match)) |>
```
Comment (maurolepore) [inline R/prepare_matches_mapper.R:17]: As I explain in another comment, if we assert that `multi_match` is logical, then we don't need this line. In that case what we do need is a test to ensure that passing `multi_match` of type "character" throws the expected error.

File: R/prepare_matches_mapper.R [hunk 10]
```diff
@@ -12,10 +12,11 @@ prepare_matches_mapper <- function(mapper, activities) {
   mapper |>
     left_join(activities, by = "activity_uuid_product_uuid") |>
     summarise(across(everything(), ~ paste0(na.omit(unique(.)), collapse = "; ")), .by = c("country", "main_activity", "clustered")) |>
-    ungroup() |>
+    mutate(across(everything(), ~ifelse(. == "", NA_character_, .))) |>
```
Comment (maurolepore) [inline R/prepare_matches_mapper.R:15]: This line seems correct but could improve a bit:

* `everything()` is misleading since we only operate on strings. Thus it's more precise to use `across(where(is.character), ...`.
* `. == ""` is clear but `nzchar()` is more focused and likely faster.
* I think the tidyverse team seems to be moving away from creating lambdas with `~` in favor or either `function(x)`, `\(x)` (since R 4.0), or just creating a named function.
``` r
library(dplyr, warn.conflicts = FALSE)

data <- tibble::tibble(x = c(TRUE, FALSE), y = c("", "a"))
data
#> # A tibble: 2 × 2
#>   x     y    
#>   <lgl> <chr>
#> 1 TRUE  ""   
#> 2 FALSE "a"

# `?nzchar`: nzchar is a fast way to find out if elements of a character vector are non-empty strings.
data |> mutate(across(where(is.character), \(x) ifelse(!nzchar(x), NA_character_, x)))
#> # A tibble: 2 × 2
#>   x     y    
#>   <lgl> <chr>
#> 1 TRUE  <NA> 
#> 2 FALSE a

replace_empty_with_na <- function(x) ifelse(!nzchar(x), NA_character_, x)
data |> mutate(across(where(is.character), replace_empty_with_na))
#> # A tibble: 2 × 2
#>   x     y    
#>   <lgl> <chr>
#> 1 TRUE  <NA> 
#> 2 FALSE a
```

<sup>Created on 2023-10-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/prepare_matches_mapper.R [hunk 11]
```diff
@@ -12,10 +12,11 @@ prepare_matches_mapper <- function(mapper, activities) {
...
     distinct() |>
+    mutate(multi_match = as.logical(.data$multi_match)) |>
     mutate(multi_match = case_when(
-      multi_match == "" ~ "FALSE",
+      is.na(multi_match) ~ FALSE,
       TRUE ~ multi_match
     ))
```
Comment (maurolepore) [inline R/prepare_matches_mapper.R:21]: I see only two cases so maybe we can simplify with `ifelse()`?

If so I think the code would be something like this:

```r
  mutate(multi_match = ifelse(is.na(multi_match), FALSE, TRUE)
```

**Review by maurolepore — COMMENTED** [review]
Thanks Kalash, see my comments.

File: tests/testthat/test-prepare_matches_mapper.R [hunk 12]
```diff
@@ -1,9 +1,36 @@
 test_that("`multi_match` column has either `TRUE` or `FALSE` values", {
   out <- prepare_matches_mapper(small_matches_mapper, ecoinvent_activities)
-  expect_equal(unique(out$multi_match), c("TRUE", "FALSE"))
+  expect_equal(unique(out$multi_match), c(TRUE, FALSE))
 })
```
[the remark this answers is not in the document]
Comment (kalashsinghal) [reply tests/testthat/test-prepare_matches_mapper.R]: @maurolepore Thanks for explaining this! I now get the gist of it why you ask me to change the titles. Next time I will keep this in mind.

File: tests/testthat/test-prepare_matches_mapper.R [hunk 13]
```diff
@@ -1,9 +1,36 @@
...
 
 test_that("`completion` column has either `low`, `medium`, or `high` values", {
   out <- prepare_matches_mapper(small_matches_mapper, ecoinvent_activities)
   expect_equal(unique(out$completion), c("low", "high", "medium"))
 })
+
+test_that("`prepare_matches_mapper` doesn't have any empty string", {
```
[the remark this answers is not in the document]
Comment (kalashsinghal) [reply tests/testthat/test-prepare_matches_mapper.R]: @kalashsinghal Understood :) Resolved

File: tests/testthat/test-prepare_matches_mapper.R (hunk 8 above)
Comment (kalashsinghal) [reply tests/testthat/test-prepare_matches_mapper.R]: @maurolepore Indeed you are right that the `multi_match` column should remain logical. However, this line of code change the type of this column:
```summarise(across(everything(), ~ paste0(na.omit(unique(.)), collapse = "; ")), .by = c("country", "main_activity", "clustered"))```

After analysing the real data, I identified that the requirement to run this command is only required for these two columns `activity_uuid_product_uuid` and `activity_name`. Hence, I have applied a new check to ensure that the need to run this command on other columns will not change in the future (by stoping the code from running). Please let me know if the new logic make sense to you! We can have a call this time to align easily if you'd like :)

File: R/prepare_matches_mapper.R (hunk 9 above)
Comment (kalashsinghal) [reply R/prepare_matches_mapper.R]: @maurolepore Added test!

File: R/prepare_matches_mapper.R (hunk 11 above)
Comment (kalashsinghal) [reply R/prepare_matches_mapper.R]: Yes make sense! changed

File: R/prepare_matches_mapper.R (hunk 10 above)
Comment (kalashsinghal) [reply R/prepare_matches_mapper.R]: @kalashsinghal we have a logical column `multi_match` which is not the character column, hence I have added `everything()` in the new commit in all the commands. Please let me know if you would like to see this change!

I have applied all the updates you recommended. Thanks for taking out so much time on this!

File: R/prepare_matches_mapper.R [hunk 14]
```diff
@@ -6,16 +6,24 @@
...
+
+  non_summarised_df |>
+    left_join(summarised_df, by = c("country", "main_activity", "clustered")) |>
+    mutate(across(everything(), \(x) ifelse(!nzchar(x), NA_character_, x))) |>
     distinct() |>
-    mutate(multi_match = case_when(
-      multi_match == "" ~ "FALSE",
-      TRUE ~ multi_match
-    ))
+    mutate(multi_match = ifelse(is.na(.data$multi_match), FALSE, TRUE))
```
Comment (maurolepore) [inline R/prepare_matches_mapper.R:28]: `tidyr::replace_na()` seems more evocative:

``` r
data <- tibble::tibble(multi_match = c(TRUE, FALSE, NA), x = c(TRUE, FALSE, NA))
data
#> # A tibble: 3 × 2
#>   multi_match x    
#>   <lgl>       <lgl>
#> 1 TRUE        TRUE 
#> 2 FALSE       FALSE
#> 3 NA          NA

tidyr::replace_na(data, list(multi_match = FALSE))
#> # A tibble: 3 × 2
#>   multi_match x    
#>   <lgl>       <lgl>
#> 1 TRUE        TRUE 
#> 2 FALSE       FALSE
#> 3 FALSE       NA
```

<sup>Created on 2023-10-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/prepare_matches_mapper.R [hunk 15]
```diff
@@ -6,16 +6,24 @@
...
+  non_summarised_df <- mapper_df |>
+    select(-c("activity_uuid_product_uuid", "activity_name")) |>
+    summarise(across(everything(), \(x) unique(x)), .by = c("country", "main_activity", "clustered"))
+  summarised_df <- mapper_df |>
+    select(c("country", "main_activity", "clustered", "activity_uuid_product_uuid", "activity_name")) |>
+    summarise(across(everything(), \(x) paste0(na.omit(unique(x)), collapse = "; ")), .by = c("country", "main_activity", "clustered"))
+
+  non_summarised_df |>
+    left_join(summarised_df, by = c("country", "main_activity", "clustered")) |>
+    mutate(across(everything(), \(x) ifelse(!nzchar(x), NA_character_, x))) |>
```
Comment (maurolepore) [inline R/prepare_matches_mapper.R:26]: This became more complicated. It's hard for me to understand what's exactly your intent but this might help:

The call to `summarize()` with `unique()` seems to do something similar to `distinct()`. So maybe something like this works for you?

```r
  # ...
  mapper_df |>
    distinct(country, main_activity, clustered) |>
    # I'm not sure about this though. Maybe this inserts the duplicates that the line above removes?
    right_join(mapper_df) |>
    #...
```

File: tests/testthat/test-prepare_matches_mapper.R [hunk 16]
```diff
@@ -1,9 +1,58 @@
...
+
+  expect_error(prepare_matches_mapper(mm, ea), "activity_uuid_product_uuid.*country")
+})
+
+test_that("stops if type of column `multi_match` is not logical", {
+  mm <- tibble_names("a", names(matches_mapper))
+  ea <- tibble_names("a", names(ecoinvent_activities))
+
+  expect_error(prepare_matches_mapper(mm, ea), "multi_match.*TRUE")
+})
```
Comment (maurolepore) [inline tests/testthat/test-prepare_matches_mapper.R:58]: We may clarify the  expected value in the input and in the expected string, e.g.:

```r
mm <- tibble_names("a", names(matches_mapper))
all_characters_none_logical <- "a"
ea <- tibble_names(all_characters_none_logical, names(ecoinvent_activities))

expect_error(prepare_matches_mapper(mm, ea), "logical.*multi_match.*not TRUE")
```

File: R/utils.R [hunk 17]
```diff
@@ -49,3 +49,22 @@ modify_col <- function(data, pattern, f) {
...
+
+  if (!all(unlist(check_col) == 1)) {
+    abort(c(
+      "Columns other than `activity_uuid_product_uuid` and `activity_name`
+      should not have more than one unique value for `country`, `main_activity`,
+      and `clustered` group of columns."
+    ))
+  }
+  invisible(matches_mapp)
+}
```
[the remark this answers is not in the document]
Comment (kalashsinghal) [reply R/utils.R]: Yes it worked! Implemented

File: R/prepare_matches_mapper.R [hunk 18]
```diff
@@ -6,16 +6,24 @@
...
   activities <- activities |>
     select("activity_uuid_product_uuid", "activity_name", "reference_product_name", "unit")
 
-  mapper |>
-    left_join(activities, by = "activity_uuid_product_uuid") |>
-    summarise(across(everything(), ~ paste0(na.omit(unique(.)), collapse = "; ")), .by = c("country", "main_activity", "clustered")) |>
-    ungroup() |>
+  mapper_df <- mapper |>
+    left_join(activities, by = "activity_uuid_product_uuid")
+  stop_if_some_columns_have_more_than_one_unique_value(mapper_df)
```
[the remark this answers is not in the document]
Comment (kalashsinghal) [reply R/prepare_matches_mapper.R]: @maurolepore I have used it in the pipe. 
I want to assert all the all columns after join, hence I have applied the stop function for `mapper_df`. Please let me know if know a better way to apply it. Thanks!

File: R/prepare_matches_mapper.R (hunk 14 above)
Comment (kalashsinghal) [reply R/prepare_matches_mapper.R]: Implemented! Thanks for suggesting improved solution

File: tests/testthat/test-prepare_matches_mapper.R (hunk 16 above)
Comment (kalashsinghal) [reply tests/testthat/test-prepare_matches_mapper.R]: Makes more sense! Applied

File: R/prepare_matches_mapper.R (hunk 15 above)
Comment (kalashsinghal) [reply R/prepare_matches_mapper.R]: @maurolepore  Thanks for suggesting this! 
However, after merging this code #67 from another PR, I found new errors. To resolve them, I have to refactor the code a lot. Please review it again! Thanks!

**Review by maurolepore — APPROVED** [review]
Thanks @kalashsinghal for your patience. 
From a code-syntax perspective it LGTM.

I still struggle to understand what the function does so here are some suggestions. In any case I understand you'll move the function to data-preparation, so you may continue there.

* Name the function more evocatively. The name "prepare" is vague and gives no hint about what to expect from the code in it's body.
* Extract each code chunk into an internal helper and give it a meaningful name. That will give you a lot of room to capture your intent inside the body of the function. It'll move away disctracting details to lower level functions, leaving just the high-level concepts that explain what's going on.
* Write more tests. Large and unclear logic is a great place for bugs to hide. Expected behaviour, edge-cases, and bad inputs to ensure you excercise as much code as possible and crush as many bugs as you can find.
